### PR TITLE
Add support for `pre-ghstack` git hook

### PR DESCRIPTION
**Pull request by amartani**
As a follow-up to #92, this allows moving checks that we
currently run on git's `pre-push` hook to be run at a `pre-ghstack`
time instead by configuring a git hook named `pre-ghstack`.
When running the submit action, if this hook is found, run it,
passing the base and head commit ids as arguments (to help
the hook to determine the changes that are being submitted),
and only proceed if successful.

For reference, our intended implementation of this hook is to run:

`pre-commit run --hook-stage push --from-ref "$1" --to-ref "$2"`

### CODE REVIEW

**Comment by amartani** [comment]
@ezyang not sure if you got notified of this.. Do you think this is reasonable? I've been using this successfully for my use case, but let me know if you prefer a different approach.

File: ghstack/submit.py [hunk 1]
```diff
@@ -1051,3 +1054,15 @@ def format_url(s: DiffMeta) -> str:
...
+def run_pre_ghstack_hook(sh: ghstack.shell.Shell, base_commit: str, top_commit: str) -> None:
+    """If a `pre-ghstack` git hook is configured, run it."""
+    default_hooks_path = os.path.join(sh.git("rev-parse", "--show-toplevel"), ".git/hooks")
+    hooks_path = sh.git("config", "--default", default_hooks_path, "--get", "core.hooksPath")
+    hook_file = os.path.join(hooks_path, "pre-ghstack")
+
+    if not os.path.isfile(hook_file) or not os.access(hook_file, os.X_OK):
+        return
+
+    sh.sh(hook_file, base_commit, top_commit, stdout=None)
```
Comment (ezyang) [inline ghstack/submit.py:1068]: I think my questions are mostly around whether or not this is the Correct(TM) way to retrieve custom hooks from Git. I'll have to check the Git source code on the matter.

Comment (amartani) [reply ghstack/submit.py:1068]: I based the approach on the implementation of `git-p4`, which is part of `git` itself: https://github.com/git/git/blob/90d242d36e248acfae0033274b524bfa55a947fd/git-p4.py#L259

I didn't do the bits that are specific for Windows (is `ghstack` supported on windows?), but other than that, this is as official as I could find.

Comment (ezyang) [reply ghstack/submit.py:1068]: Hook implementation in git-p4.py: https://github.com/git/git/blob/89bece5c8c96f0b962cfc89e63f82d603fd60bed/git-p4.py#L224-L288

Comment (ezyang) [reply ghstack/submit.py:1068]: and the C implementation https://github.com/git/git/blob/89bece5c8c96f0b962cfc89e63f82d603fd60bed/hook.c

Comment (ezyang) [reply ghstack/submit.py:1068]: oops I missed your 3hr comment

**Review by ezyang — APPROVED** [review]

**Comment by amartani** [comment]
@ezyang would you be able to publish a new version of ghstack to pypi with this change? Thank you!

**Pull request merged**